### PR TITLE
Added byteswap (from C++23)

### PR DESCRIPTION
**Pull request by Chi-Iroh**
Hello :wave: 

First of all, I really like the project, and I'd like to help a little when I have the time to.
I added byteswap function (from \<bit\> header since C++23), and I used __bit/bit_cast.h as a reference to be consistent.

I have no warning and all tests (bit_14, 17, 20 and 23) pass.
Please tell me if more tests are needed, I used values from cppreference (values were slightly modified for signed integers to avoid overflow).

### CODE REVIEW

**Comment by lackhole** [comment]
Hello @Chi-Iroh 👋

Thank you so much for your contribution! I'm glad to hear that you like the project, and I really appreciate you taking the time to help out. Your addition of the byteswap function looks great, and it's reassuring to know that all the tests are passing.

I just need to figure out how to run GitHub Actions on the repository before merging your PR. Once I have that set up, I'll proceed with merging your changes.

Please feel free to contribute anytime you have the time—your help is always welcome!

**Review by lackhole — CHANGES_REQUESTED** [review]
Hi @Chi-Iroh. Thank you for your contribution! 
I double-checked your code since I couldn't toggle GitHub Action manually and I found some changes to be applied.

And could you update your branch with `main`(rebase or merge, any form)? I made the changes to the `main` so that I can toggle Action manually but it doesn't seem to work until your branch have the latest changes to the `main`.

Whether it is small or big, any contribution are really helpful to me. So don't hesitate to make changes :)

----------

### `T` should model `integral`. 
  * Using `static_assert` allows `byteswap` to participate in the overload resolution.
  * See https://eel.is/c++draft/bit.byteswap#1 or https://en.cppreference.com/w/cpp/numeric/byteswap for the definition
  * See https://en.cppreference.com/w/cpp/language/sfinae#Library_support for using SFINAE

### `UnsignedT byteswap_impl(T n)`
  * `byteswap` must return `T`
  * Invoking function must specify it's full namespaces to prevent ADL and some compilation error on old compilers

### Notes about `bytemask`
  * Use `unsigned char` to reduce memory usage

### Notes about dispatching under `namespace detail`(A.K.A impl)
  * I prefer using [tag dispatching](https://www.fluentcpp.com/2018/04/27/tag-dispatching/)(e.g., [addressof.h](https://github.com/lackhole/stl-preview/blob/main/include/preview/__memory/addressof.h#L47-L55)) than SFINAE in `detail` if possible. Because evaluating SFINAE usually takes more compilation time than tag dispatching.

### Notes about overflow
  * Overflow happens iif `sizeof(T) == 1`, but in such case the return value is unchanged. So it will be better to check the size at compile time, and use faster method(avoid multiplication -`*`- which is too slow for the purpose of this function)

### Notes about formatting
Although I didn't mentioned it anywhere, I follow the [Google C++ styleguide](https://google.github.io/styleguide/cppguide.html) for _formatting_. So there should be 
  * no indents inside the namespaces
  * namespace closing should always be commented
  * (You can just follow the formatting of `bit_cast.h`)

**Comment by Chi-Iroh** [comment]
I just updated my _byteswap_ branch to your main, so you should be able to run your actions now.

> Using `static_assert` allows `byteswap` to participate in the overload resolution.

Oops, I saw that but forgot it and put the `static_assert` anyway to give an explicit message. I'll remove it.

> `byteswap` must return `T`

Obviously. I forgot to convert it back to a signed type after calling `byteswap_impl<UnsignedT>`.

> Invoking function must specify it's full namespaces to prevent ADL and some compilation error on old compilers

Honestly, I hadn't thought about that as I have never had to deal with this kind of errors, but that's good to know.

> Notes about dispatching under `namespace detail` (A.K.A impl)

No problem, thank you for the link about tag dispatching, I'll see what it is.

> Notes about formatting

Alright. It would be relevant to mention the styleguide in the README to inform potential contributors.


Thank you for taking the time to explain what I did wrong.

**Comment by Chi-Iroh** [comment]
I'll reopen the PR once all the changes you mentioned will be done.

**Comment by Chi-Iroh** [comment]
I think it's better now, but feel free to correct me if I'm still wrong.
I didn't make `byte_mask` unsigned char, because `unsigned char << unsigned` yields an int, so `byte_mask << upper_shift` would result in an int and `static_cast<T>(byte_mask)` would be needed to avoid overflow.
Without adding the static_cast, 64 bits tests fail.

**Comment by lackhole** [comment]
@Chi-Iroh 
Ternary operator _can_ be evaluated at compile-time, but it is not a compile-time dispatching. Try something like below to get more optimized and compile-time dispatching code:)

```cpp
template<typename T, std::size_t Size>
constexpr T byteswap_unsigned_impl(T n, std::integral_constant<Size>) { return /* generic method */; }
template<typename T>
constexpr T byteswap_unsigned_impl(T n, std::integral_constant<1>) { return n; /* unchanged */ }
template<typename T>
constexpr T byteswap_unsigned_impl(T n, std::integral_constant<2>) { return ...; /* specialized method */ }
// ... more specialization for Size, stop at 32 because 64 and 128 is not always present

template<typename T>
constexpr T byteswap_unsigned(T n) { return byteswap_unsigned_impl(n, std::integral_constant<sizeof(T)>{}); }
```

+ Note: explicit template parameter is not required when calling `byteswap_impl`.

**Comment by Chi-Iroh** [comment]
Thanks for the example, I understand now.
My commits are marked as unverified, don't worry, it's just because I didn't import my GPG key on my Windows for now.

**Comment by lackhole** [comment]
It seems like clang tester is somehow not working. I'll merge after fixing it

**Review by lackhole — APPROVED** [review]

**Comment by Chi-Iroh** [comment]
Nice, so I guess the code suits you better now ?

**Pull request merged**
